### DISCUSSION TITLE
Easy way to extend the Request class?

### DISCUSSION THREAD

**Discussion by gthedev** [laravel/framework#38650] [Q&A] [answered]
Is there an easy way to extend the default `Illuminate\Http\Request` class? Across multiple projects I have found that it would be useful to have a custom class that can automatically be injected as the default `Request` class but extend it with some extra methods.

For example, the `ApiRequest` class could extend the `Request` class to have methods to check the versioning based on the headers or query params. Or in another case with multiple different authenticatable user Models the extended Request class could have different methods for `->user()` `->customer()` `->admin()` etc.

Currently to achieve that this is what I did in my `AppServiceProvider`:
```
        $this->app->bind(ApiRequest::class, function () {
            $request = ApiRequest::createFrom($this->app['request']);

            return $request;
        });
```

Feels a little hacky and additionally getting `app('request')` will still return the original `Request` class instance. Is there a better way to achieve this and extend/override the default `Request` class?

**Comment by fun4life1990** [comment]
Hi!

You can create your own descendant of Laravel Request. For example:

```
class BaseRequest extends Request
{
    public function expectsJson()
    {
        return true;
    }

    public function wantsJson()
    {
        return true;
    }
}
```

And then use it directly in your index.php:

```
$kernel = $app->make(Kernel::class);

$response = tap($kernel->handle(
    $request = BaseRequest::capture()
))->send();

$kernel->terminate($request, $response);
```

And that all!

However in your artisan you would still get Laravel native request.

**Comment by dennisprudlo** [comment, marked as answer]
The default request class is macroable.
https://github.com/laravel/framework/blob/eb378fc5f4b32b73caec0977aae169079523fe98/src/Illuminate/Http/Request.php#L20-L25

So you could simply extend the request class. Put this in the boot method of one of your service providers:
```php
Request::macro('apiVersion', function ()  {
    return request()->header('api-version') ?? 'latest';
});
```

And then in your controller or wherever you need the api version:
```php
$apiVersion = request()->apiVersion();
```

**Reply by gthedev** [reply]
Thanks Dennis, that is what I do often now, my only problem is that there is no type-hinting/static-analysis support for this as the IDE & analysis tools do not understand this magic 😞

**Reply by dennisprudlo** [reply]
Yeah thats true. Another way would be to overwrite the `request` item in the container.
```php
// for dependency injection
$this->app->bind(ApiRequest::class, function () {
    return ApiRequest::createFrom($this->app['request']);
});

// for the global helpers to retrieve the request
$this->app->bind('request', function () {
    return $this->app[ApiRequest::class];
});

app('request') // ApiRequest
request() // ApiRequest
```

But in any way the `request()` helper type-hints a return value of `Illuminate\Http\Request` so you will need to overwrite that type-hint so your IDE will know it's a different type:

```php
/** @var ApiRequest */
request()->...
```

Or of course you use dependency injection

**Reply by gthedev** [reply]
Yeah your first code snippet is exactly what I ended up doing together with the Dependency Injection to get the proper hinting & analysis, I guess then that's the best way to achieve this at the moment. Thanks!
